### PR TITLE
Tool restriction

### LINKED ISSUE #6: Restrict lumberjack tool usage depending on his level
- [x] Level 1 stone
- [ ] Level 2 iron
- [ ] Level 3 diamond

### PR DESCRIPTION
**Pull request by MarioFernandes73**
adapted checkfortool / checkforpickaxe / getmostefficienttool to include tools restrictions
added new methods to handle item selection from the inventory

Closes #6 

# Changes proposed in this pull request:
-adaptability of checkfor methods to ensure the worker says an acceptable sentece
-adaptability of getmosteffiecienttool so that for every item he iterates through he will check if the item is the correct tool for the job and if it that tool has the appropriate level
-new method verifyToolLevel to check if a certain item has a certain level
-new method hasToolLevel which calls verifyToolLevel to check if a worker has at least one good item to do the job in his inventory

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 1]
```diff
@@ -1092,19 +1110,101 @@ private int getMostEfficientTool(@NotNull final Block target)
...
+                {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+    
+    
+     /**
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java]: MINOR First sentence should end with a period. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 2]
```diff
@@ -1092,19 +1110,101 @@ private int getMostEfficientTool(@NotNull final Block target)
...
+    {
+        final int hutLevel = worker.getWorkBuilding().getBuildingLevel();
+        final boolean itemEnchanted = item.hasEffect();
+
+        switch (hutLevel)
+        {
+            case 0:
+            case 1:
+            case 2:
+            {
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java]: MAJOR Extract this nested code block into a method. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1199)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 3]
```diff
@@ -1092,19 +1110,101 @@ private int getMostEfficientTool(@NotNull final Block target)
...
+            case 4:
+            {
+                if (hutLevel > level)
+                {
+                    return true;
+                }
+            }
+            break;
+            default:
+            {
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java]: MAJOR Extract this nested code block into a method. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1199)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 4]
```diff
@@ -1092,19 +1110,101 @@ private int getMostEfficientTool(@NotNull final Block target)
...
+                {
+                    bestSlot = i;
+                    bestLevel = level;
+                }
             }
         }
         return bestSlot;
     }
+    
+     /**
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java]: MINOR First sentence should end with a period. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 5]
```diff
@@ -1092,19 +1110,101 @@ private int getMostEfficientTool(@NotNull final Block target)
...
+    
+    
+     /**
+     * Verifies if an item has an appropriated grade
+     *
+     * @param item  the type of tool needed
+     * @param level the type of tool needed
+     * @return true if tool is acceptable
+     */
+    private boolean verifyToolLevel(final ItemStack item, int level)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java]: MAJOR The Cyclomatic Complexity of this method "verifyToolLevel" is 12 which is greater than 10 authorized. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 6]
```diff
@@ -1092,19 +1110,101 @@ private int getMostEfficientTool(@NotNull final Block target)
...
+            case 1:
+            case 2:
+            {
+                if (hutLevel >= level && !itemEnchanted)
+                {
+                    return true;
+                }
+            }
+            break;
+            case 3:
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java]: MINOR Assign this magic number 3 to a well-named constant, and use the constant instead. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 7]
```diff
@@ -1092,19 +1110,101 @@ private int getMostEfficientTool(@NotNull final Block target)
         final int required = target.getHarvestLevel(target.getDefaultState());
         int bestSlot = -1;
         int bestLevel = Integer.MAX_VALUE;
-        @NotNull final InventoryCitizen inventory = worker.getInventoryCitizen();
+        @NotNull InventoryCitizen inventory = worker.getInventoryCitizen();
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java]: MINOR Local variable 'inventory' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 8]
```diff
@@ -1092,19 +1110,101 @@ private int getMostEfficientTool(@NotNull final Block target)
...
+            case 2:
+            {
+                if (hutLevel >= level && !itemEnchanted)
+                {
+                    return true;
+                }
+            }
+            break;
+            case 3:
+            case 4:
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java]: MINOR Assign this magic number 4 to a well-named constant, and use the constant instead. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 9]
```diff
@@ -1092,19 +1110,101 @@ private int getMostEfficientTool(@NotNull final Block target)
...
+     /**
+     * Verifies if there is one tool with an acceptable level
+     * in a worker's inventory
+     *
+     * @param tool the type of tool needed
+     * @return true if tool is acceptable
+     */
+    private boolean hasToolLevel(final String tool)
+    {
+        @NotNull InventoryCitizen inventory = worker.getInventoryCitizen();
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java]: MINOR Local variable 'inventory' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 11 issues
* MAJOR 5 major
* MINOR 6 minor

Watch the comments in this conversation to review them.

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 10]
```diff
@@ -1092,19 +1110,101 @@ private int getMostEfficientTool(@NotNull final Block target)
...
+            {
+                if (hutLevel > level)
+                {
+                    return true;
+                }
+            }
+            break;
+            default:
+            {
+                return true;
```
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java]: So items with a level > 4 are okay to be used at whatever level?

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 11]
```diff
@@ -1092,19 +1110,101 @@ private int getMostEfficientTool(@NotNull final Block target)
...
+     * @param item  the type of tool needed
+     * @param level the type of tool needed
+     * @return true if tool is acceptable
+     */
+    private boolean verifyToolLevel(final ItemStack item, int level)
+    {
+        final int hutLevel = worker.getWorkBuilding().getBuildingLevel();
+        final boolean itemEnchanted = item.hasEffect();
+
+        switch (hutLevel)
```
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java]: I'd suggest converting this switch into if/else cases, they are way clearer for this.

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java (hunk 5 above)
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java]: I would say we should move this method into utils to not bloat the ai code too much

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 12]
```diff
@@ -1092,19 +1110,101 @@ private int getMostEfficientTool(@NotNull final Block target)
...
     }
+    
+     /**
+     * Verifies if there is one tool with an acceptable level
+     * in a worker's inventory
+     *
+     * @param tool the type of tool needed
+     * @return true if tool is acceptable
+     */
+    private boolean hasToolLevel(final String tool)
```
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java]: This method could also be moved to utils

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 13]
```diff
@@ -1092,19 +1110,101 @@ private int getMostEfficientTool(@NotNull final Block target)
...
+        for (int i = 0; i < inventory.getSizeInventory(); i++)
+        {
+            final ItemStack item = inventory.getStackInSlot(i);
+            final int level = Utils.getMiningLevel(item, tool);
+
+            if (item != null)
+            {
+                final boolean isTool = Utils.isTool(item, tool);
+                final boolean isUsable = verifyToolLevel(item, level);
+                if (isTool && isUsable)
```
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java]: no need to have two temporaty variables if you only use them in this one if, just inline them

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java (hunk 7 above)
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java]: Why the change?

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 14]
```diff
@@ -1092,19 +1110,101 @@ private int getMostEfficientTool(@NotNull final Block target)
...
             {
-                bestSlot = i;
-                bestLevel = level;
+                boolean isUsable = true;
+
+                if (item != null)
+                {
+                    isUsable = verifyToolLevel(item, level);
+                }
+                if (isUsable)
```
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java]: Like already commented in the commit, this will be true for all items that are null and that may be not good

Comment (marvin-bitterlich) [reply src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java]: Best would be to say: 
`if (item != null && verifyToolLevel(item, level)`

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 15]
```diff
@@ -627,7 +637,7 @@ private AIState waitForPickaxe()
         return IDLE;
     }
 
-    /**
+   /**
```
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java]: This format is wrong, the space is intended there ;) Have you imported the intellij formatter?

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 16]
```diff
@@ -519,15 +519,25 @@ private boolean checkForShovel()
...
                                        worker.getInventoryCitizen(),
                                        stack -> Utils.isTool(stack, tool),
                                        InventoryFunctions::doNothing
                                      );
-        if (!needsTool)
+
+        final boolean isUsable = hasToolLevel(tool);
+        final int hutLevel = worker.getWorkBuilding().getBuildingLevel();
+
+        if (!needsTool && isUsable)
```
Comment (marvin-bitterlich) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:542]: why &&
if he does not need a tool, break
if the tool he got is fine, break

If I am not mistaken, a || will work and allow you to inline isUseable

**Review by marvin-bitterlich — CHANGES_REQUESTED** [review]
The switch case should definitely be refactored and we may be able to draw some methods into utils

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 17]
```diff
@@ -658,11 +675,12 @@ private boolean checkForPickaxe(final int minlevel)
             {
                 return true;
             }
-            requestWithoutSpam("Pickaxe at least level " + minlevel);
+            requestWithoutSpam("Pickaxe of at least level " + minlevel + " and at maximum " + hutLevel);
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java]: Can we put this in our languageHandler and lang files so we're able to translate it?

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java (hunk 14 above)
Comment (marvin-bitterlich) [reply src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java]: Okay I got it now! You need to pull the nullcheck into verifyToolLevel. It should definitely be @Nullable and return true when the item is null

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java (hunk 11 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java]: Agree. The switch case is hard to interpret in this case.

**Review by Raycoms — CHANGES_REQUESTED** [review]

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java (hunk 16 above)
Comment (MarioFernandes73) [reply src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:542]: the needsTool doesn't take into consideration the level, so, a worker with level 0 hut "doesn't need a shovel" if he already has a diamond shovel in his inventory, but since it's a diamond one, he can't use it, therefore we need those 2 to be verified

Comment (marvin-bitterlich) [reply src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:542]: That seems weird to me. Can we not just do the isUsable check inside the calculation for needsTool? That way the logic is where it belongs

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 18]
```diff
@@ -1086,23 +1107,30 @@ private void updateToolFlag(@NotNull final String tool, final int required)
...
         for (int i = 0; i < inventory.getSizeInventory(); i++)
         {
             final ItemStack item = inventory.getStackInSlot(i);
             final int level = Utils.getMiningLevel(item, tool);
+
             if (level >= required && level < bestLevel)
             {
-                bestSlot = i;
-                bestLevel = level;
+                if (tool == null || Utils.verifyToolLevel(item, level, hutLevel))
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java]: MAJOR Merge this if statement with the enclosing one. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1066)

File: src/main/java/com/minecolonies/coremod/util/Utils.java [hunk 19]
```diff
@@ -434,4 +438,55 @@ public static int getFortuneOf(@Nullable final ItemStack tool)
...
+
+            if (Utils.isTool(item, tool) && verifyToolLevel(item, level, hutLevel))
+            {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/util/Utils.java]: MINOR First sentence should end with a period. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)

File: src/main/java/com/minecolonies/coremod/util/Utils.java [hunk 20]
```diff
@@ -434,4 +438,55 @@ public static int getFortuneOf(@Nullable final ItemStack tool)
...
+        if (item == null || hutLevel > 4)
+        {
+            return true;
+        }
+        else if (item.hasEffect() && hutLevel <= 2 )
+        {
+            return false;
+        }
+        else if (hutLevel >= level)
+        {
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/util/Utils.java]: MAJOR This branch's code block is the same as the block for the branch on line 477. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1871)

File: src/main/java/com/minecolonies/coremod/util/Utils.java [hunk 21]
```diff
@@ -1,13 +1,17 @@
 package com.minecolonies.coremod.util;
 
 import com.minecolonies.compatibility.Compatibility;
+import com.minecolonies.coremod.entity.EntityCitizen;
+import com.minecolonies.coremod.inventory.InventoryCitizen;
 import net.minecraft.block.Block;
+import net.minecraft.block.SoundType;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/util/Utils.java]: MINOR Remove this unused import 'net.minecraft.block.SoundType'. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/minecolonies/coremod/util/Utils.java [hunk 22]
```diff
@@ -434,4 +438,55 @@ public static int getFortuneOf(@Nullable final ItemStack tool)
         }
         return fortune;
     }
+    
+    /**
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/util/Utils.java]: MINOR First sentence should end with a period. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)

File: src/main/java/com/minecolonies/coremod/util/Utils.java [hunk 23]
```diff
@@ -434,4 +438,55 @@ public static int getFortuneOf(@Nullable final ItemStack tool)
...
+     * Verifies if an item has an appropriated grade
+     *
+     * @param item   the type of tool needed
+     * @param level  the type of tool needed
+     * @param hutLevel the worker's hut level
+     * @return true if tool is acceptable
+     */
+    public static boolean verifyToolLevel(final ItemStack item, int level, final int hutLevel)
+    {
+        if (item == null || hutLevel > 4)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/util/Utils.java]: MINOR Assign this magic number 4 to a well-named constant, and use the constant instead. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/coremod/util/Utils.java [hunk 24]
```diff
@@ -1,13 +1,17 @@
...
+import com.minecolonies.coremod.inventory.InventoryCitizen;
 import net.minecraft.block.Block;
+import net.minecraft.block.SoundType;
 import net.minecraft.init.Blocks;
 import net.minecraft.item.ItemHoe;
 import net.minecraft.item.ItemStack;
 import net.minecraft.item.ItemSword;
 import net.minecraft.item.ItemTool;
 import net.minecraft.nbt.NBTTagList;
+import net.minecraft.util.SoundCategory;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/util/Utils.java]: MINOR Remove this unused import 'net.minecraft.util.SoundCategory'. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 8 issues
* MAJOR 2 major
* MINOR 6 minor

Watch the comments in this conversation to review them.

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 5 issues
* MAJOR 2 major
* MINOR 3 minor

Watch the comments in this conversation to review them.

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 25]
```diff
@@ -540,7 +552,7 @@ private boolean checkForTool(@NotNull final String tool)
         {
             return false;
         }
-        requestWithoutSpam(tool);
+        chatSpamFilter.talkWithoutSpam(LanguageHandler.format("entity.worker.toolRequest",tool,hutLevel));
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java]: can we put a space after each "," ?

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 26]
```diff
@@ -658,11 +678,12 @@ private boolean checkForPickaxe(final int minlevel)
             {
                 return true;
             }
-            requestWithoutSpam("Pickaxe at least level " + minlevel);
+            chatSpamFilter.talkWithoutSpam(LanguageHandler.format("entity.worker.pickaxeRequest",minlevel,hutLevel));
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java]: same thing here.

File: src/main/java/com/minecolonies/coremod/util/Utils.java (hunk 22 above)
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/util/Utils.java]: Please end the sentence with a period. I once worked hours to fill them all in.

Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/util/Utils.java]: And I feel like this does fit more in "inventoryUtils"

File: src/main/java/com/minecolonies/coremod/util/Utils.java (hunk 19 above)
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/util/Utils.java]: same thing here.

File: src/main/java/com/minecolonies/coremod/util/Utils.java (hunk 23 above)
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/util/Utils.java]: Please make it something like "FREE_TOOL_CHOICE_LEVEL"

File: src/main/java/com/minecolonies/coremod/util/Utils.java (hunk 20 above)
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/util/Utils.java]: can't we merge all true statements in the same if?

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 4 issues
* MAJOR 2 major
* MINOR 2 minor

Watch the comments in this conversation to review them.

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java [hunk 27]
```diff
@@ -1086,23 +1107,30 @@ private void updateToolFlag(@NotNull final String tool, final int required)
...
         for (int i = 0; i < inventory.getSizeInventory(); i++)
         {
             final ItemStack item = inventory.getStackInSlot(i);
             final int level = Utils.getMiningLevel(item, tool);
+
             if (level >= required && level < bestLevel)
             {
-                bestSlot = i;
-                bestLevel = level;
+                if (tool == null || InventoryUtils.verifyToolLevel(item, level, hutLevel))
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java:1126]: MAJOR Merge this if statement with the enclosing one. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1066)

File: src/main/java/com/minecolonies/coremod/util/Utils.java [hunk 28]
```diff
@@ -1,6 +1,7 @@
 package com.minecolonies.coremod.util;
 
 import com.minecolonies.compatibility.Compatibility;
+import com.minecolonies.coremod.inventory.InventoryCitizen;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/util/Utils.java]: MINOR Remove this unused import 'com.minecolonies.coremod.inventory.InventoryCitizen'. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/minecolonies/coremod/util/InventoryUtils.java [hunk 29]
```diff
@@ -622,4 +627,55 @@ private static int findSlotForItemStack(@NotNull final IInventory inventory, @No
...
+
+            if (Utils.isTool(item, tool) && verifyToolLevel(item, level, hutLevel))
+            {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/util/InventoryUtils.java]: MINOR First sentence should end with a period. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)

File: src/main/java/com/minecolonies/coremod/util/InventoryUtils.java [hunk 30]
```diff
@@ -622,4 +627,55 @@ private static int findSlotForItemStack(@NotNull final IInventory inventory, @No
...
+        if (item == null || hutLevel > FREE_TOOL_CHOICE_LEVEL)
+        {
+            return true;
+        }
+        else if (item.hasEffect() && hutLevel <= EFFECT_TOOL_CHOICE_LEVEL )
+        {
+            return false;
+        }
+        else if (hutLevel >= level)
+        {
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/util/InventoryUtils.java:674]: MAJOR This branch's code block is the same as the block for the branch on line 666. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1871)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 4 issues
* MAJOR 2 major
* MINOR 2 minor

Watch the comments in this conversation to review them.

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 3 issues
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: src/main/java/com/minecolonies/coremod/util/InventoryUtils.java [hunk 31]
```diff
@@ -622,4 +627,74 @@ private static int findSlotForItemStack(@NotNull final IInventory inventory, @No
...
+     * @return a string corresponding to the tool
+     */
+    public static String swapToolGrade(final int toolGrade)
+    {
+        switch(toolGrade)
+        {
+            case 0: return "Wood or Gold";
+            case 1: return "Stone";
+            case 2: return "Iron";
+            case 3: return "Diamond";
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/util/InventoryUtils.java]: MINOR Assign this magic number 3 to a well-named constant, and use the constant instead. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS109)

**Review by Raycoms — APPROVED** [review]

**Comment by Raycoms** [comment]
@Kostronor can you look at the javadoc build. Something is going wrong there. But Isn't related to their changes.

**Review by marvin-bitterlich — APPROVED** [review]

**Comment by marvin-bitterlich** [comment]
:shipit: @Raycoms 

[Approved with PullApprove](https://pullapprove.com/Minecolonies/minecolonies/pull-request/352/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)

**Comment by Raycoms** [comment]
:shipit: 

[Approved with PullApprove](https://pullapprove.com/Minecolonies/minecolonies/pull-request/352/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)

File: src/main/java/com/minecolonies/coremod/util/InventoryUtils.java [hunk 32]
```diff
@@ -622,4 +627,77 @@ private static int findSlotForItemStack(@NotNull final IInventory inventory, @No
...
+    {
+        switch (toolGrade)
+        {
+            case 0:
+                return "Wood or Gold";
+            case 1:
+                return "Stone";
+            case 2:
+                return "Iron";
+            case 3:
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/util/InventoryUtils.java:697]: MINOR Assign this magic number 3 to a well-named constant, and use the constant instead. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS109)
